From: Jiri Kosina <jkosina@suse.cz>
Subject: [PATCH] floppy: do not corrupt bio.bi_flags when reading block 0
References: bnc#879258
Patch-mainline: v3.16-rc1
Git-commit: 6314a108ec19aefa5160535b2bfe1ca9c38efe37

Commit 41a55b4de39 ("floppy: silence warning during disk test") caused
bio.bi_flags being overwritten, and its initialization to BIO_UPTODATE
in bio_init() to be lost.

This was unnoticed until 7b7b68bba5 ("floppy: bail out in open() if
drive is not responding to block0 read"), because the error value wasn't
checked for in the bio completion callback.

Now we are actually looking at the error, and the loss of BIO_UPTODATE
causes EIO to be wrongly passed to the callback, which confuses the
FD_OPEN_SHOULD_FAIL_BIT logic.

Fix this by not destroying previous value of bi_flags when setting
BIO_QUIET.

Signed-off-by: Jiri Kosina <jkosina@suse.cz>
---
 drivers/block/floppy.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-3.12-SLE12/drivers/block/floppy.c
===================================================================
--- linux-3.12-SLE12.orig/drivers/block/floppy.c	2014-05-28 15:09:29.733948298 +0200
+++ linux-3.12-SLE12/drivers/block/floppy.c	2014-05-28 15:10:17.068210400 +0200
@@ -3798,7 +3798,7 @@
 	bio.bi_size = size;
 	bio.bi_bdev = bdev;
 	bio.bi_sector = 0;
-	bio.bi_flags = (1 << BIO_QUIET);
+	bio.bi_flags |= (1 << BIO_QUIET);
 	bio.bi_private = &cbdata;
 	bio.bi_end_io = floppy_rb0_cb;
 
